packages/data/src/components/use-select/index.js

```diff
@@ -104,7 +105,100 @@ const renderQueue = createQueue();
  * ```
  * @return {UseSelectReturn<T>} A custom react hook.
  */
+function Selecter( registry ) {
+	const NOTHING = {};
+	let lastMapSelect = NOTHING;
+	let lastMapResult = NOTHING;
+	let lastMapResultValid = false;
+
+	const NULLSCRIBER = () => {
+		lastMapResultValid = false;
+		return () => {};
+	};
+	let subscribe = NULLSCRIBER;
+
+	function createSubscriber( stores ) {
+		if ( stores.length === 0 ) {
+			return NULLSCRIBER;
+		}
+
+		return ( lis ) => {
+			lastMapResultValid = false;
+			const unsubs = stores.map( ( storeName ) => {
+				return registry.subscribe( () => {
+					lastMapResultValid = false;
+					lis();
+				}, storeName );
+			} );
+
+			return () => {
+				// The return value of the subscribe function could be undefined if the store is a custom generic store.
+				unsubs.forEach( ( unsub ) => unsub?.() );
+			};
+		};
+	}
+
+	function select( mapSelect, resub ) {
+		function getValue() {
+			if ( lastMapResultValid && mapSelect === lastMapSelect ) {
+				return lastMapResult;
+			}
+
+			const mapResult = mapSelect( registry.select, registry );
+
+			if (
+				lastMapResult === NOTHING ||
+				! isShallowEqual( lastMapResult, mapResult )
+			) {
+				lastMapResult = mapResult;
+			}
+			lastMapResultValid = true;
+
+			return lastMapResult;
+		}
+
+		if ( lastMapResultValid && mapSelect === lastMapSelect ) {
+			return { getValue, subscribe };
+		}
+
+		const listeningStores = { current: null };
+		const mapResult = registry.__unstableMarkListeningStores(
+			() => mapSelect( registry.select, registry ),
+			listeningStores
+		);
+
+		if (
+			lastMapSelect === NOTHING ||
+			( resub && mapSelect !== lastMapSelect )
+		) {
+			subscribe = createSubscriber( listeningStores.current );
+		}
+
+		if (
+			lastMapResult === NOTHING ||
+			! isShallowEqual( lastMapResult, mapResult )
+		) {
+			lastMapResult = mapResult;
+		}
+		lastMapResultValid = true;
+
+		lastMapSelect = mapSelect;
+
+		return { getValue, subscribe };
+	}
+
+	return { select };
+}
+
 export default function useSelect( mapSelect, deps ) {
+	const registry = useRegistry();
+	const selecter = useMemo( () => Selecter( registry ), [ registry ] );
+	const selector = useCallback( mapSelect, deps );
+	const { getValue, subscribe } = selecter.select( selector, !! deps );
```

Comment:
Do we need to be a bit smarter about whether the `deps` changed? Won't this always be truthy when any dependencies are provided?

Reply:
The main place where `deps` are used is the `useCallback` hook that creates a memoized version of `mapSelect`. `useCallback` with no deps is an identity function, which returns the callback without memoizing. The `!! deps` boolean is used to distinguish between two `useSelect` modes:

1. Without deps. That's what mainly the `withSelect` HOC does. In that case, the hook only subscribes to the stores once, on mount, and doesn't resubscribe when `mapSelect` changes. `mapSelect` can be a different function on every call, and the latest version will always be used.
2. With deps. In that case, the hook resubscribes to stores on every `deps` change. And it can also avoid calls to `mapSelect` if the last value is valid and memoized `mapSelect` hasn't changed. See the condition at the beginning of `updateValue`. Therefore, `deps` let us run through the `useSelect` hook really fast if nothing relevant has changed.

I added a lot of comments to the new hook implementation to explain all the details.